### ISSUE TITLE
Unpack1FFT broken

### ISSUE THREAD

**Issue by dkmayer** [supercollider/supercollider#3214]
This is probably related to #3212. The help file example in an executable form (see # 3213) work in 3.6.6, but not since 3.7.

(
s.boot.doWhenBooted {
    var fftsize = 1024;
    b = Buffer.alloc(s, fftsize, 1);
    c = Buffer.read(s, Platform.resourceDir +/+ "sounds/a11wlk01.wav");
}
)

// Let's extract the DC component - i.e. the magnitude at index zero.
(
x = {
    var sig, chain, unp;
    sig = PlayBuf.ar(1, c, BufRateScale.kr(c), loop: 1);
    chain = FFT(b, sig);

    unp = Unpack1FFT(chain, b.numFrames, 0, 0);

    // Demand some data from the unpacker
    Demand.kr(chain>=0, 0, unp).poll(chain>=0, "unpacked value");

    (sig*0.1).dup;
}.play(s);
)

**Comment by patrickdupuis** [comment]
This issue is unclear. Are you saying that the example from the `Unpack1FFT` Help is broken, or the Ugen itself? It seems to me from testing that the UGen is not broken, just the example from the Help. Also confusing is that the above code snipet is not exactly the same as the Help example. The code below is taken from Help :

```supercollider
(
s.waitForBoot({
    c = Buffer.read(s, Platform.resourceDir +/+ "sounds/a11wlk01.wav");
})
)

// Let's extract the DC component - i.e. the magnitude at index zero.
(
x = {
    var fftsize = 1024;
    var sig, chain, unp;
    sig = PlayBuf.ar(1, c, BufRateScale.kr(c), loop: 1);
    chain = FFT(LocalBuf(fftsize), sig);

    unp = Unpack1FFT(chain, b.numFrames, 0, 0);

    // Demand some data from the unpacker
    Demand.kr(chain>=0, 0, unp).poll(chain>=0, "unpacked value");

    (sig*0.1).dup;
}.play(s);
)
x.free;
```

This example does not function due to the fact that the variable `b` is not defined anywhere. When `b.numFrames` is replaced by `1024` it works fine. Note that in this example, `FFT` is using a `LocalBuf` and not buffer `b`.

This issue is a duplicate of #3213 and should be closed.

**Comment by dkmayer** [comment]
>This issue is unclear. Are you saying that the example from the Unpack1FFT Help is broken, or >the Ugen itself?

The latter, therefore it's different from #3213 (but might be the same source as #3212). Indeed you are right that I was not duplicating exactly the help file, but that doesn't change the fact that my above posted example works in 3.6.6. and not later (or does it post for you in newer builds ?). Couldn't the combination of our observations that the example works with LocalBuf and it doesn't with an externally defined buffer give a clue, what's happening here ?

Also the example from http://new-supercollider-mailing-lists-forums-use-these.2681727.n2.nabble.com/FFT-beginnings-td7627987.html
works with the version of SynthDef \fft_trigger which is using Unpack1FFT in 3.6.6 and not later.
But this is admittedly a complicated reproducer, however there are at least two on my OSX machine.

**Comment by patrickdupuis** [comment]
I notice that the original posted example works if you change the rate of polling to something slower (like the default value 10). The issue isn't related to the externally defined `Buffer` then. 

Interestingly, I am able run it correctly if i add a `trig` variable like so:
```supercollider
(
s.boot.doWhenBooted {
    var fftsize = 1024;
    b = Buffer.alloc(s, fftsize, 1);
    c = Buffer.read(s, Platform.resourceDir +/+ "sounds/a11wlk01.wav");
}
)

(
x = {
    var sig, chain, unp, trig;

    sig = PlayBuf.ar(1, c, BufRateScale.kr(c), loop: 1);
    chain = FFT(b, sig);
    unp = Unpack1FFT(chain, b.numFrames, 0, 0);
    trig = chain>=0;
    Demand.kr(trig, 0, unp).poll(trig, "unpacked value");

    (sig*0.1).dup;
}.play(s);
)
```

I guess this means that what we are dealing with here is a case where the compiled code running on the server isn't quite what we expect. Assigning the trigger to a variable seems to put things in the correct order. Someone with more knowledge than me can hopefully explain this more clearly.

**Comment by telephon** [comment]
Normally, a trigger is `trig = chain>0;` and not `trig = chain>=0;`. But I don't know if that changes anything.

**Comment by jamshark70** [comment]
The FFT bufnum may be 0; so FFT triggers are -1 for no calculation and >= 0 to calculate.

**Comment by telephon** [comment]
sorry, I couldn't find information about this: what has the trigger signal to do with the buffer number?

**Comment by jamshark70** [comment]
The `chain` input to PV UGens is not only a trigger. It's also the buffer index.

If it were not the buffer index, how would the unit know which buffer to operate on?

**Comment by telephon** [comment]
I know, but always assumed it returns some kind of array.

**Comment by jamshark70** [comment]
Nope. It's just a signal -- when it's time for an FFT frame, the FFT unit outputs the buffer number; when it isn't, it's -1. Downstream PV units use this to know when to fire and where to look for FFT data.

(But of course this isn't in the help: we only document FFT.new, whose output is, according to FFT help, "the FFT chain"... ???)

**Comment by telephon** [comment]
*That* would be a good bit of documentation indeed. I'll try and add it.

**Comment by telephon** [comment]
How to demonstrate this?

I tried 

```supercollider

(
{ 
	var chain = FFT(LocalBuf(128), DC.ar()); 
	var trig = chain >= 0; 
	IFFT.ar(chain);
	[chain, Trig.ar(trig, 0.001)]
}.plot(0.01)
)
```

but I don't get anything

**Comment by telephon** [comment]
Also I just saw that `PV_ChainUGen::addCopiesIfNeeded` is not documented anywhere.

**Comment by jamshark70** [comment]
`chain` is kr, not ar, so you can't plot it at audio rate.

```
(
{ 
	var chain = FFT(LocalBuf(512), DC.ar());
	var trig = chain >= 0;
	var min = RunningMin.kr(chain), max = RunningMax.kr(chain);
	[min, max].poll(Impulse.kr(1, 0.9));  // should wait 10th of a second
	chain
}.plot(128 * 64/44100)
)

->
UGen Array [1]: 2049
UGen Array [0]: -1   <<-- there it is
```

**Comment by jamshark70** [comment]
Also, if you use a 128-sample FFT buffer, and your hop size is 0.5, then there will be an FFT frame every 64 samples, and FFT will output the buffer number continuously, so your plot wouldn't show the triggering behavior.

(For FFT, the normal trigger rule about the transition from <= 0 to > 0 doesn't apply -- and IMO the documentation should be careful not to call it a trigger, because it isn't exactly a trigger.)

**Comment by telephon** [comment]
thanks, @jamshark70.

We can plot kr signals: `{ SinOsc.kr(50) }.plot(1/50)`, but indeed the hop sizes make it different.

**Comment by telephon** [comment]
I found this, added it to help:

```supercollider
// the FFT return signal is -1, and for each starting window,
// it is the FFT buffer number

b = Buffer.alloc(s, 512); // allocate FFT buffer
b.bufnum; // note the buffer number

(
var dt = s.options.blockSize / s.sampleRate * 16; // plot 16 blocks
var min = -2, max = b.bufnum + 4;
// input is SoundIn, but we don't see this signal
{ FFT(b, SoundIn.ar) }.plot(dt, minval:min, maxval:max).plotMode_(\steps);
)
```

**Comment by snappizz** [comment]
is this resolved?

**Comment by brianlheim** [comment]
@telephon @dkmayer - what's the status of this issue?

**Comment by dkmayer** [comment]
Hi Brian, our summer term starts tomorrow, so I'm sorry to have no time to look at this at the moment. My remembrance is that there were still open questions.
